src/components/Form/CreateProfileForm/DiscountData/Bucket.tsx

```diff
@@ -68,56 +64,79 @@ Props) => {
-            setUploadProgress(Math.round((100 * event.loaded) / event.total));
-          }
-        }),
-      toError
-    )
-      .chain(chainAxios)
-      .map(response => response.data)
-      .fold(
-        () => {
-          setUploadingDoc(false);
-          setUploadProgress(0);
-          triggerTooltip({
-            severity: Severity.DANGER,
-            text: "Caricamento del file fallito"
-          });
-        },
-        data => {
-          setFieldValue(
-            hasIndex
-              ? `discounts[${index}].lastBucketCodeLoadUid`
-              : "lastBucketCodeLoadUid",
-            data.uid
-          );
-          setFieldValue(
-            hasIndex
-              ? `discounts[${index}].lastBucketCodeLoadFileName`
-              : "lastBucketCodeLoadFileName",
-            files[0].name
-          );
-          setCurrentDoc({ name: files[0].name });
-          setUploadingDoc(false);
-          setUploadProgress(0);
-          setCanUploadFile(false);
+    const response = await normalizeAxiosResponse(
+      Api.Bucket.uploadBucket(agreementId, files[0], {
+        onUploadProgress: (event: any) => {
+          setUploadProgress(Math.round((100 * event.loaded) / event.total));
         }
-      )
-      .run();
+      })
+    );
+    if (response.status === 200 || response.status === 204) {
+      setFieldValue(
+        hasIndex
+          ? `discounts[${index}].lastBucketCodeLoadUid`
+          : "lastBucketCodeLoadUid",
+        response.data.uid
+      );
+      setFieldValue(
+        hasIndex
+          ? `discounts[${index}].lastBucketCodeLoadFileName`
+          : "lastBucketCodeLoadFileName",
+        files[0].name
+      );
+      setCurrentDoc({ name: files[0].name });
+      setCanUploadFile(false);
+    } else if (
+      response.status === 400 &&
+      response.data === "CSV_NAME_OR_EXTENSION_NOT_VALID"
+    ) {
+      triggerTooltip({
+        severity: Severity.DANGER,
+        text:
+          "Il formato del documento non è valido. Carica un documento CSV e riprova."
+      });
+    } else if (
+      response.status === 400 &&
+      response.data === "MAX_ALLOWED_BUCKET_CODE_LENGTH_NOT_RESPECTED"
+    ) {
+      triggerTooltip({
+        severity: Severity.DANGER,
+        text: "Ogni codice della lista non deve superare i 20 caratteri"
+      });
+    } else if (
+      response.status === 400 &&
+      response.data ===
+        "BUCKET_CODES_MUST_BE_ALPHANUM_WITH_AT_LEAST_ONE_DIGIT_AND_CHAR"
+    ) {
+      triggerTooltip({
+        severity: Severity.DANGER,
+        text:
+          "Ogni codice della lista deve avere almeno un numero e una lettera."
+      });
+    } else if (
+      response.status === 400 &&
+      response.data === "CANNOT_LOAD_BUCKET_FOR_NOT_RESPECTED_MINIMUM_BOUND"
+    ) {
+      triggerTooltip({
+        severity: Severity.DANGER,
+        text:
+          "La lista caricata deve contenere almeno 10000 codici. Carica un'altra lista e riprova."
+      });
```

Comment:
I see here a lot of code duplicated, what do you think to handle the error codes that has the status `400` with a function something like the following:

```typescript
// Outside the component declare a map with the messages
const ERROR_MESSAGES = {
  CSV_NAME_OR_EXTENSION_NOT_VALID: "Il formato del documento non è valido. Carica un documento CSV e riprova.",
  MAX_ALLOWED_BUCKET_CODE_LENGTH_NOT_RESPECTED: "Ogni codice della lista non deve superare i 20 caratteri.",
  BUCKET_CODES_MUST_BE_ALPHANUM_WITH_AT_LEAST_ONE_DIGIT_AND_CHAR: "Ogni codice della lista deve avere almeno un numero e una lettera.",
  CANNOT_LOAD_BUCKET_FOR_NOT_RESPECTED_MINIMUM_BOUND: "La lista caricata deve contenere almeno 10000 codici. Carica un'altra lista e riprova.",
  DEFAULT: "Caricamento del file fallito."
};

// Then create a function that trigger the tooltip with the message
const handleErrorMessages = (errorCode: string) => {
  const message = ERROR_MESSAGES[errorCode] || ERROR_MESSAGES.DEFAULT;
  triggerTooltip({ severity: Severity.DANGER, text: message });
};
```

In this way, we can reduce the code duplicated and it can be more maintainable in the future:

```typescript
const addFile = async (files: any) => {
    setUploadingDoc(true);
    const response = await normalizeAxiosResponse(
      Api.Bucket.uploadBucket(agreementId, files[0], {
        onUploadProgress: (event: any) => {
          setUploadProgress(Math.round((100 * event.loaded) / event.total));
        }
      })
    );
    if (response.status === 200 || response.status === 204) {
      setFieldValue(
        hasIndex
          ? discounts[${index}].lastBucketCodeLoadUid
          : "lastBucketCodeLoadUid",
        response.data.uid
      );
      setFieldValue(
        hasIndex
          ? discounts[${index}].lastBucketCodeLoadFileName
          : "lastBucketCodeLoadFileName",
        files[0].name
      );
      setCurrentDoc({ name: files[0].name });
      setCanUploadFile(false);
    } else if (response.status === 400) {
      handleErrorMessages(response.data);
    } else {
      triggerTooltip({
        severity: Severity.DANGER,
        text: "Caricamento del file fallito"
      });
    }
    setUploadingDoc(false);
    setUploadProgress(0);
  };
```

Reply:
ok, for me both options are the same